lighthouse-core/report/html/renderer/report-ui-features.js

```diff
@@ -130,6 +133,76 @@ class ReportUIFeatures {
     dropdown.addEventListener('click', this.onExport);
   }
 
+  _setupThirdPartyFilter() {
+    // get all tables with a text url
+    /** @type {Array<HTMLTableElement>} */
+    const tables = Array.from(this._document.querySelectorAll('.lh-table'));
+    const thirdPartyFilterAuditExclusions = tables
+      .filter(el => el.querySelector('td.lh-table-column--url'))
+      .filter(el => /** @type {Element} */ (el.closest('.lh-audit')).id !== 'uses-rel-preconnect');
+
+    thirdPartyFilterAuditExclusions.forEach((tableEl, index) => {
+      /** @type {Record<string, HTMLTableRowElement>} */
+      const thirdPartyRows = this._getThirdPartyRows(tableEl, this.json.finalUrl);
+
+      // create input box
+      const filterTemplate = this._dom.cloneTemplate('#tmpl-lh-3p-filter', this._document);
+      const filterInput = /** @type {HTMLInputElement} */ (filterTemplate.querySelector('input'));
+      const id = `lh-3p-filter-label--${index}`;
+
+      filterInput.setAttribute('id', id);
+      filterInput.addEventListener('change', e => {
+        // remove elements from the dom and keep track of them to readd on uncheck
+        // why removing instead of hiding? nth-child(even) background-colors keep working
+        if (e.target instanceof HTMLInputElement && e.target.checked) {
+          Object.entries(thirdPartyRows).forEach(([position, row]) => {
+            const childrenArr = Array.from(tableEl.tBodies[0].children);
+            tableEl.tBodies[0].insertBefore(row, childrenArr[Number(position)]);
+          });
+        } else {
+          Object.keys(thirdPartyRows).forEach(position => {
+            const row = thirdPartyRows[position];
+            row.remove();
+          });
+        }
+      });
+
+      if (tableEl.parentNode) {
+        /** @type {Element} */ (filterTemplate.querySelector('label')).setAttribute('for', id);
+        /** @type {Element} */ (filterTemplate.querySelector('.lh-3p-filter-count')).textContent =
+          `${Object.keys(thirdPartyRows).length}`;
+        // Finally, add checkbox to the DOM
+        tableEl.parentNode.insertBefore(filterTemplate, tableEl);
+      }
+    });
+  }
+
+  /**
+   * @param {HTMLTableElement} el
+   * @param {string} finalUrl
+   */
+  _getThirdPartyRows(el, finalUrl) {
+    /** @type {NodeListOf<HTMLElement>} */
+    const urlItems = el.querySelectorAll('.lh-text__url');
+    const pageTLDPlusOne = new URL(finalUrl).origin.split('.').slice(-2).join('.');
+    /** @type {Record<string, HTMLTableRowElement>} */
+    const thirdPartyRows = {};
```

Comment:
feels like a good use of Map or WeakMap instead... just to communicate it's holding elements rather than data.